### PR TITLE
Switch testrender to cone tracing

### PR DESCRIPTION
**Pull request by aconty**
## Description

The idea is to stop propagating derivs through BSDFs, which we
never did properly. Instead we trace rays with radius + spread,
which we turn into OSL derivs in SG when we hit a surface. Then
when bouncing off a surface we use the BSDF roughness to set the
spread. In addition we also:

  * Return BSDF samples/eval as a small struct Sample(). This
    avoids output parameters in the methods.
  * Stop passing a full ShaderGlobals to the BSDF methods. Pass
    a viewing direction wo instead so they could run in a simpler
    context.

## Tests

Updated testrender tests because of very minimal diffs.

## Checklist:

- [x] I have read the [contribution guidelines](../CONTRIBUTING.md).
- [x] I have previously submitted a [Contributor License Agreement](http://opensource.imageworks.com/cla/).
- [x] I have updated the documentation, if applicable.
- [x] I have ensured that the change is tested somewhere in the testsuite (adding new test cases if necessary).
- [x] My code follows the prevailing code style of this project.

### CODE REVIEW

File: src/testrender/shading.cpp [hunk 1]
```diff
@@ -972,165 +949,168 @@ struct MxBurleyDiffuse final : public BSDF, MxBurleyDiffuseParams {
...
 
 
-Color3 clamp(const Color3& c, float min, float max)
+Color3
+clamp(const Color3& c, float min, float max)
 {
-    return Color3(c.x < min ? min : max < c.x ? max : c.x,
-		  c.y < min ? min : max < c.y ? max : c.y,
-		  c.z < min ? min : max < c.z ? max : c.z);
+    return Color3(c.x < min   ? min
```
Comment (fpsunflower) [inline src/testrender/shading.cpp:1014]: This line looks like it got formatted a bit strangely?

**Review by fpsunflower — APPROVED** [review]
LGTM!

There are a few spots where it looks like clang-format had too narrow of a width that you might want to double check ...

**Comment by aconty** [comment]
Maybe, but that's the project configuration as far as I know.

**Review by fpsunflower — APPROVED** [review]

**Review by fpsunflower — APPROVED** [review]
LGTM

Sorry about the merge conflicts 😅

**Comment by aconty** [comment]
No problem! Thanks

**Pull request merged**